### PR TITLE
Adding data elements to experiments

### PR DESCRIPTION
**Pull request by jandom**

### CODE REVIEW

**Comment by jandom** [comment]
@BarneyLabstep this is ready for another review

File: labstep/protocol.py [hunk 1]
```diff
@@ -327,6 +329,77 @@ def getSteps(self):
...
+            dataElement = protocol.addDataElement(fieldName="Refractive Index",
+                                               value="1.73")
+        """
+        metadataThread = self.last_version['metadata_thread']
+        filterParams = {'metadata_thread_id': metadataThread['id'],
+                        'type': fieldType,
+                        'label': fieldName,
+                        'value': value,
+                        'date': handleDate(date),
+                        'quantity_amount': quantity_amount,
```
Comment (BarneyLabstep) [inline labstep/protocol.py:398]: Does this actually work? I suspect not as these fields have been changed to `number` and `unit` in the db. 

If so let's see a test covering it!

If not, let's find and replace to `number` and `unit` everywhere (including metadata elsewhere). This naming makes more sense anyway.

File: labstep/protocol.py [hunk 2]
```diff
@@ -327,6 +329,77 @@ def getSteps(self):
...
+            protocol = user.getProtocol(17000)
+            metadata = exp_protocol.getDataElements()
+        """
+        metadataThread = self.last_version['metadata_thread']
+        return listToClass(
+            metadataThread['metadatas'],
+            Metadata, self.__user__
+        )
+
+    def addDataElement(self, fieldType="default", fieldName=None,
```
Comment (BarneyLabstep) [inline labstep/protocol.py:355]: This parameter is required by the backend so let's take away the 'None' default to make it required here (you'll also have to move it to be the first param)

File: labstep/protocol.py (hunk 1 above)
Comment (jandom) [reply labstep/protocol.py:398]: sure I'm happy to do it but it falls outside of the core of this PR – it's clearly related but will increase scope

**Comment by jandom** [comment]
@BarneyLabstep , ha – finally green – see if this is good now?

File: labstep/resourceItem.py [hunk 3]
```diff
@@ -103,7 +103,7 @@ def newResourceItem(user, resource, name=None, availability=None,
 
 
 def editResourceItem(resourceItem, name=None, availability=None,
-                     quantity_amount=None, quantity_unit=None,
+                     number=None, unit=None,
```
Comment (BarneyLabstep) [inline labstep/resourceItem.py:106]: Sorry for resource_item the backend does still use `quantity_amount` and `quantity_unit`!

Comment (jandom) [reply labstep/resourceItem.py:106]: Good catch!

File: labstep/resource.py [hunk 4]
```diff
@@ -164,29 +164,29 @@ def delete(self):
...
-        fieldName (str)
-            The name of the field.
         value (str)
             The value accompanying the fieldName entry.
         date (str)
             The date and time accompanying the fieldName entry. Must be
             in the format of "YYYY-MM-DD HH:MM".
-        quantity_amount (float)
+        number (float)
             The quantity.
```
Comment (BarneyLabstep) [inline labstep/resource.py:187]: Change this to 'The numeric value for metadata of type number.'

File: labstep/resource.py [hunk 5]
```diff
@@ -318,9 +318,9 @@ def newItem(self, name=None, availability='available',
         availability (str)
             The status of the ResourceItem. Options are:
             "available" and "unavailable".
-        quantity_amount (float)
+        number (float)
```
Comment (BarneyLabstep) [inline labstep/resource.py:321]: revert

File: labstep/resource.py [hunk 6]
```diff
@@ -318,9 +318,9 @@ def newItem(self, name=None, availability='available',
...
             The quantity of the ResourceItem.
-        quantity_unit (str)
+        unit (str)
```
Comment (BarneyLabstep) [inline labstep/resource.py:323]: revert

File: labstep/resource.py [hunk 7]
```diff
@@ -338,6 +338,6 @@ def newItem(self, name=None, availability='available',
             item = my_resource.newItem(name='Test Item')
         """
         return newResourceItem(self.__user__, self, name,
-                               availability, quantity_amount,
-                               quantity_unit, location,
+                               availability, number,
```
Comment (BarneyLabstep) [inline labstep/resource.py:341]: revert

File: labstep/resourceItem.py [hunk 8]
```diff
@@ -60,7 +60,7 @@ def getResourceItems(user, resource, count=100, search_query=None,
 
 
 def newResourceItem(user, resource, name=None, availability=None,
-                    quantity_amount=None, quantity_unit=None,
+                    number=None, unit=None,
```
Comment (BarneyLabstep) [inline labstep/resourceItem.py:63]: revert

File: labstep/resourceItem.py [hunk 9]
```diff
@@ -77,9 +77,9 @@ def newResourceItem(user, resource, name=None, availability=None,
     availability (str)
         The status of the ResourceItem. Options are:
         "available" and "unavailable".
-    quantity_amount (float)
+    number (float)
```
Comment (BarneyLabstep) [inline labstep/resourceItem.py:80]: revert

File: labstep/resourceItem.py [hunk 10]
```diff
@@ -77,9 +77,9 @@ def newResourceItem(user, resource, name=None, availability=None,
...
         The quantity of the ResourceItem.
-    quantity_unit (str)
+    unit (str)
```
Comment (BarneyLabstep) [inline labstep/resourceItem.py:82]: revert

File: labstep/resourceItem.py [hunk 11]
```diff
@@ -92,8 +92,8 @@ def newResourceItem(user, resource, name=None, availability=None,
     filterParams = {'resource_id': resource.id,
                     'name': name,
                     'status': handleString(availability),
-                    'quantity_amount': quantity_amount,
-                    'quantity_unit': quantity_unit}
+                    'number': number,
```
Comment (BarneyLabstep) [inline labstep/resourceItem.py:95]: revert

File: labstep/resourceItem.py [hunk 12]
```diff
@@ -117,9 +117,9 @@ def editResourceItem(resourceItem, name=None, availability=None,
     availability (str)
         The status of the ResourceItem. Options are:
         "available" and "unavailable".
-    quantity_amount (float)
+    number (float)
```
Comment (BarneyLabstep) [inline labstep/resourceItem.py:120]: revert

File: labstep/resourceItem.py [hunk 13]
```diff
@@ -117,9 +117,9 @@ def editResourceItem(resourceItem, name=None, availability=None,
...
         The quantity of the ResourceItem.
-    quantity_unit (str)
+    unit (str)
```
Comment (BarneyLabstep) [inline labstep/resourceItem.py:122]: revert

File: labstep/resourceItem.py [hunk 14]
```diff
@@ -133,11 +133,11 @@ def editResourceItem(resourceItem, name=None, availability=None,
     """
     filterParams = {'name': name,
                     'status': handleString(availability),
-                    'quantity_unit': quantity_unit,
+                    'unit': unit,
```
Comment (BarneyLabstep) [inline labstep/resourceItem.py:136]: revert

File: labstep/resourceItem.py [hunk 15]
```diff
@@ -133,11 +133,11 @@ def editResourceItem(resourceItem, name=None, availability=None,
...
                     'deleted_at': deleted_at}
 
-    if quantity_amount is not None:
-        filterParams['quantity_amount'] = float(quantity_amount)
+    if number is not None:
+        filterParams['number'] = float(number)
```
Comment (BarneyLabstep) [inline labstep/resourceItem.py:140]: revert

File: labstep/resourceItem.py [hunk 16]
```diff
@@ -162,7 +162,7 @@ class ResourceItem(Entity):
     __entityName__ = 'resource-item'
 
     def edit(self, name=None, availability=None,
-             quantity_amount=None, quantity_unit=None,
+             number=None, unit=None,
```
Comment (BarneyLabstep) [inline labstep/resourceItem.py:165]: revert

File: labstep/resourceItem.py [hunk 17]
```diff
@@ -174,9 +174,9 @@ def edit(self, name=None, availability=None,
         availability (str)
             The status of the OrderRequest. Options are:
             "available" and "unavailable".
-        quantity_amount (float)
+        number (float)
```
Comment (BarneyLabstep) [inline labstep/resourceItem.py:177]: revert

File: labstep/resourceItem.py [hunk 18]
```diff
@@ -174,9 +174,9 @@ def edit(self, name=None, availability=None,
...
             The quantity of the ResourceItem.
-        quantity_unit (str)
+        unit (str)
```
Comment (BarneyLabstep) [inline labstep/resourceItem.py:179]: revert

File: labstep/resourceItem.py [hunk 19]
```diff
@@ -194,7 +194,7 @@ def edit(self, name=None, availability=None,
             my_resource_item.edit(name='A New ResourceItem Name')
         """
         return editResourceItem(self, name, availability,
-                                quantity_amount, quantity_unit,
+                                number, unit,
```
Comment (BarneyLabstep) [inline labstep/resourceItem.py:197]: revert

**Comment by jandom** [comment]
No, i'm not reverting these changes – this API is a mess. If you want to, do so yourself

**Comment by jandom** [comment]
@BarneyLabstep I'm assigning this PR to you – you seem to have an understanding of this madness that I lack.

**Pull request merged**